File: server-php/src/Service/AdminService.php
<?php

declare(strict_types=1);

namespace SmartClock\Server\Service;

use PDO;
use SmartClock\Server\Support\ApiException;
use SmartClock\Server\Support\Time;

final class AdminService
{
    public function __construct(
        private readonly PDO $pdo,
        private readonly SyncService $syncService
    ) {
    }

    public function listUsers(?string $query, int $limit): array
    {
        $limit = max(1, min($limit, 200));
        if ($query === null || trim($query) === '') {
            $stmt = $this->pdo->prepare(
                "SELECT id, phone, email, nickname, role, status FROM users ORDER BY id DESC LIMIT $limit"
            );
            $stmt->execute();
        } else {
            $keyword = '%' . trim($query) . '%';
            $stmt = $this->pdo->prepare(
                "SELECT id, phone, email, nickname, role, status
                 FROM users
                 WHERE phone LIKE :kw OR email LIKE :kw OR nickname LIKE :kw
                 ORDER BY id DESC
                 LIMIT $limit"
            );
            $stmt->execute([':kw' => $keyword]);
        }

        return array_map([$this, 'mapUserDto'], $stmt->fetchAll());
    }

    public function getUser(int $userId): array
    {
        $stmt = $this->pdo->prepare(
            'SELECT id, phone, email, nickname, role, status FROM users WHERE id = :id LIMIT 1'
        );
        $stmt->execute([':id' => $userId]);
        $record = $stmt->fetch();
        if (!$record) {
            throw new ApiException(404, 'user not found', 'not_found');
        }

        return $this->mapUserDto($record);
    }

    public function updateUserStatus(int $adminUserId, int $userId, array $payload, string $ipAddress): array
    {
        if (!array_key_exists('status', $payload)) {
            throw new ApiException(400, 'status is required', 'bad_request');
        }
        $status = (int) $payload['status'];

        $stmt = $this->pdo->prepare(
            'UPDATE users SET status = :status, updated_at = :updated_at WHERE id = :id'
        );
        $stmt->execute([
            ':status' => $status,
            ':updated_at' => Time::toDb(Time::now()),
            ':id' => $userId,
        ]);

        $this->logAudit($adminUserId, 'USER_STATUS_UPDATE', 'USER', (string) $userId, ['status' => $status], $ipAddress);
        return $this->getUser($userId);
    }

    public function listAlarms(?int $userId, ?string $query, int $limit): array
    {
        $limit = max(1, min($limit, 300));

        if ($userId !== null && ($query === null || trim($query) === '')) {
            $stmt = $this->pdo->prepare(
                "SELECT * FROM alarms WHERE user_id = :user_id ORDER BY updated_at DESC LIMIT $limit"
            );
            $stmt->execute([':user_id' => $userId]);
        } elseif ($userId !== null) {
            $keyword = '%' . trim((string) $query) . '%';
            $stmt = $this->pdo->prepare(
                "SELECT * FROM alarms
                 WHERE user_id = :user_id AND (title LIKE :kw OR client_uuid LIKE :kw)
                 ORDER BY updated_at DESC
                 LIMIT $limit"
            );
            $stmt->execute([':user_id' => $userId, ':kw' => $keyword]);
        } elseif ($query === null || trim($query) === '') {
            $stmt = $this->pdo->prepare(
                "SELECT * FROM alarms ORDER BY updated_at DESC LIMIT $limit"
            );
            $stmt->execute();
        } else {
            $keyword = '%' . trim($query) . '%';
            $stmt = $this->pdo->prepare(
                "SELECT * FROM alarms
                 WHERE title LIKE :kw OR client_uuid LIKE :kw
                 ORDER BY updated_at DESC
                 LIMIT $limit"
            );
            $stmt->execute([':kw' => $keyword]);
        }

        return array_map(fn (array $row) => $this->syncService->mapAlarm($row), $stmt->fetchAll());
    }

    public function getAlarm(int $alarmId): array
    {
        return $this->syncService->mapAlarm($this->syncService->getAlarmById($alarmId));
    }

    public function updateAlarm(int $adminUserId, int $alarmId, array $payload, string $ipAddress): array
    {
        $existing = $this->syncService->getAlarmById($alarmId);
        $now = Time::toDb(Time::now());

        $stmt = $this->pdo->prepare(
            'UPDATE alarms SET
              title = :title, note = :note, enabled = :enabled, status = :status, trigger_time = :trigger_time,
              start_date = :start_date, end_date = :end_date, schedule_mode = :schedule_mode, alert_policy = :alert_policy,
              time_anchor_mode = :time_anchor_mode, interval_months = :interval_months, interval_years = :interval_years,
              template_id = :template_id, updated_at = :updated_at
             WHERE id = :id'
        );

        $stmt->execute([
            ':title' => array_key_exists('title', $payload) && $payload['title'] !== null
                ? $this->limitText((string) $payload['title'], 128, true)
                : $existing['title'],
            ':note' => array_key_exists('note', $payload) && $payload['note'] !== null ? (string) $payload['note'] : $existing['note'],
            ':enabled' => array_key_exists('enabled', $payload) && $payload['enabled'] !== null ? ((bool) $payload['enabled'] ? 1 : 0) : (int) $existing['enabled'],
            ':status' => array_key_exists('status', $payload) && $payload['status'] !== null ? (int) $payload['status'] : (int) $existing['status'],
            ':trigger_time' => array_key_exists('triggerTime', $payload) && $payload['triggerTime'] !== null ? Time::toDb(Time::parseIso((string) $payload['triggerTime'])) : $existing['trigger_time'],
            ':start_date' => array_key_exists('startDate', $payload) && $payload['startDate'] !== null ? Time::toDb(Time::parseIso((string) $payload['startDate'])) : $existing['start_date'],
            ':end_date' => array_key_exists('endDate', $payload) && $payload['endDate'] !== null ? Time::toDb(Time::parseIso((string) $payload['endDate'])) : $existing['end_date'],
            ':schedule_mode' => array_key_exists('scheduleMode', $payload) && $payload['scheduleMode'] !== null ? (int) $payload['scheduleMode'] : (int) $existing['schedule_mode'],
            ':alert_policy' => array_key_exists('alertPolicy', $payload) && $payload['alertPolicy'] !== null ? (int) $payload['alertPolicy'] : (int) $existing['alert_policy'],
            ':time_anchor_mode' => array_key_exists('timeAnchorMode', $payload) && $payload['timeAnchorMode'] !== null ? (int) $payload['timeAnchorMode'] : (int) $existing['time_anchor_mode'],
            ':interval_months' => array_key_exists('intervalMonths', $payload) && $payload['intervalMonths'] !== null ? max(1, (int) $payload['intervalMonths']) : (int) $existing['interval_months'],
            ':interval_years' => array_key_exists('intervalYears', $payload) && $payload['intervalYears'] !== null ? max(1, (int) $payload['intervalYears']) : (int) $existing['interval_years'],
            ':template_id' => array_key_exists('templateId', $payload) && $payload['templateId'] !== null
                ? $this->limitText((string) $payload['templateId'], 64, true)
                : $existing['template_id'],
            ':updated_at' => $now,
            ':id' => $alarmId,
        ]);

        $this->logAudit($adminUserId, 'ALARM_UPDATE', 'ALARM', (string) $alarmId, $payload, $ipAddress);
        return $this->getAlarm($alarmId);
    }

    public function softDeleteAlarm(int $adminUserId, int $alarmId, string $ipAddress): array
    {
        $stmt = $this->pdo->prepare(
            'UPDATE alarms SET status = 1, updated_at = :updated_at WHERE id = :id'
        );
        $stmt->execute([
            ':updated_at' => Time::toDb(Time::now()),
            ':id' => $alarmId,
        ]);

        $this->logAudit($adminUserId, 'ALARM_SOFT_DELETE', 'ALARM', (string) $alarmId, ['status' => 1], $ipAddress);
        return ['message' => 'ok'];
    }

    public function listAlarmLogs(?int $userId, int $limit): array
    {
        $limit = max(1, min($limit, 500));
        if ($userId === null) {
            $stmt = $this->pdo->prepare(
                "SELECT * FROM alarm_logs ORDER BY fired_at DESC LIMIT $limit"
            );
            $stmt->execute();
        } else {
            $stmt = $this->pdo->prepare(
                "SELECT * FROM alarm_logs WHERE user_id = :user_id ORDER BY fired_at DESC LIMIT $limit"
            );
            $stmt->execute([':user_id' => $userId]);
        }

        return array_map(function (array $row): array {
            return [
                'id' => (int) $row['id'],
                'alarmId' => (int) $row['alarm_id'],
                'userId' => (int) $row['user_id'],
                'firedAt' => Time::toIso(Time::parseDb($row['fired_at'] ?? null)),
                'action' => (int) $row['action'],
                'deviceId' => (string) $row['device_id'],
                'logHash' => (string) $row['log_hash'],
            ];
        }, $stmt->fetchAll());
    }

    public function listAuditLogs(int $limit): array
    {
        $limit = max(1, min($limit, 500));
        $stmt = $this->pdo->prepare(
            "SELECT * FROM admin_audit_logs ORDER BY created_at DESC LIMIT $limit"
        );
        $stmt->execute();

        return array_map(function (array $row): array {
            return [
                'id' => (int) $row['id'],
                'adminUserId' => (int) $row['admin_user_id'],
                'action' => (string) $row['action'],
                'targetType' => (string) $row['target_type'],
                'targetId' => $row['target_id'] !== null ? (string) $row['target_id'] : null,
                'detailJson' => $row['detail_json'] !== null ? (string) $row['detail_json'] : null,
                'ipAddress' => $row['ip_address'] !== null ? (string) $row['ip_address'] : null,
                'createdAt' => Time::toIso(Time::parseDb($row['created_at'] ?? null)),
            ];
        }, $stmt->fetchAll());
    }

    private function logAudit(int $adminUserId, string $action, string $targetType, ?string $targetId, mixed $detail, string $ipAddress): void
    {
        $stmt = $this->pdo->prepare(
            'INSERT INTO admin_audit_logs
             (admin_user_id, action, target_type, target_id, detail_json, ip_address, created_at)
             VALUES (:admin_user_id, :action, :target_type, :target_id, :detail_json, :ip_address, :created_at)'
        );
        $stmt->execute([
            ':admin_user_id' => $adminUserId,
            ':action' => $action,
            ':target_type' => $targetType,
            ':target_id' => $targetId,
            ':detail_json' => json_encode($detail, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES),
            ':ip_address' => $ipAddress,
            ':created_at' => Time::toDb(Time::now()),
        ]);
    }

    private function mapUserDto(array $record): array
    {
        return [
            'id' => (int) $record['id'],
            'phone' => $record['phone'] !== null ? (string) $record['phone'] : null,
            'email' => $record['email'] !== null ? (string) $record['email'] : null,
            'nickname' => $record['nickname'] !== null ? (string) $record['nickname'] : null,
            'role' => (string) $record['role'],
            'status' => (int) $record['status'],
        ];
    }

    private function limitText(?string $value, int $maxLength, bool $trim = false): ?string
    {
        if ($value === null) {
            return null;
        }

        $normalized = $trim ? trim($value) : $value;
        if ($normalized === '') {
            return $normalized;
        }

        if (function_exists('mb_substr')) {
            return mb_substr($normalized, 0, $maxLength, 'UTF-8');
        }

        if (preg_match_all('/./us', $normalized, $matches) === 1) {
            return implode('', array_slice($matches[0], 0, $maxLength));
        }

        return substr($normalized, 0, $maxLength);
    }
}

File: server/src/main/kotlin/com/smartclock/server/service/AdminService.kt
package com.smartclock.server.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.smartclock.server.dto.AdminAuditLogDto
import com.smartclock.server.dto.AlarmLogAdminDto
import com.smartclock.server.dto.AlarmSyncDto
import com.smartclock.server.dto.ApiUserDto
import com.smartclock.server.dto.PageResponse
import com.smartclock.server.dto.UpdateAlarmRequest
import com.smartclock.server.dto.toDto
import com.smartclock.server.model.UserProfile
import com.smartclock.server.support.ApiException
import com.smartclock.server.support.toInstantUtc
import com.smartclock.server.support.toIsoUtc
import com.smartclock.server.support.toSqlTimestamp
import java.sql.Timestamp
import java.time.Clock
import java.time.Instant
import org.jooq.DSLContext
import org.jooq.Record
import org.jooq.impl.DSL
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class AdminService(
    private val dsl: DSLContext,
    private val authService: AuthService,
    private val syncService: SyncService,
    private val objectMapper: ObjectMapper,
    private val clock: Clock
) {

    fun listUsers(query: String?, page: Int, pageSize: Int): PageResponse<ApiUserDto> {
        val normalizedPage = page.coerceAtLeast(1)
        val normalizedPageSize = pageSize.coerceIn(10, 100)
        val offset = (normalizedPage - 1) * normalizedPageSize
        val records = if (query.isNullOrBlank()) {
            dsl.fetch(
                "SELECT id, phone, email, nickname, role, status FROM users ORDER BY id DESC LIMIT ? OFFSET ?",
                normalizedPageSize,
                offset
            )
        } else {
            val keyword = "%${query.trim()}%"
            dsl.fetch(
                """
                SELECT id, phone, email, nickname, role, status
                FROM users
                WHERE phone LIKE ? OR email LIKE ? OR nickname LIKE ?
                ORDER BY id DESC
                LIMIT ? OFFSET ?
                """.trimIndent(),
                keyword,
                keyword,
                keyword,
                normalizedPageSize,
                offset
            )
        }
        val total = if (query.isNullOrBlank()) {
            (dsl.fetchValue("SELECT COUNT(*) FROM users") as? Number)?.toInt() ?: 0
        } else {
            val keyword = "%${query.trim()}%"
            (dsl.fetchValue(
                """
                SELECT COUNT(*)
                FROM users
                WHERE phone LIKE ? OR email LIKE ? OR nickname LIKE ?
                """.trimIndent(),
                keyword,
                keyword,
                keyword
            ) as? Number)?.toInt() ?: 0
        }
        return PageResponse(
            items = records.map(::mapUser).map(UserProfile::toDto),
            total = total,
            page = normalizedPage,
            pageSize = normalizedPageSize
        )
    }

    fun getUser(userId: Long): ApiUserDto =
        authService.me(com.smartclock.server.security.AuthUserPrincipal(userId, com.smartclock.server.model.Role.USER)).toDto()

    fun updateUserStatus(adminUserId: Long, userId: Long, status: Int, ipAddress: String): ApiUserDto = dsl.transactionResult { cfg ->
        val tx = DSL.using(cfg)
        tx.execute(
            "UPDATE users SET status = ?, updated_at = ? WHERE id = ?",
            status,
            Instant.now(clock).toSqlTimestamp(),
            userId
        )
        logAudit(tx, adminUserId, "USER_STATUS_UPDATE", "USER", userId.toString(), mapOf("status" to status), ipAddress)
        tx.fetchOne("SELECT id, phone, email, nickname, role, status FROM users WHERE id = ?", userId)
            ?.let(::mapUser)
            ?.toDto()
            ?: throw ApiException(HttpStatus.NOT_FOUND, "user not found")
    }

    fun listAlarms(userId: Long?, query: String?, page: Int, pageSize: Int): PageResponse<AlarmSyncDto> {
        val normalizedPage = page.coerceAtLeast(1)
        val normalizedPageSize = pageSize.coerceIn(10, 100)
        val offset = (normalizedPage - 1) * normalizedPageSize
        val items = when {
            userId != null && query.isNullOrBlank() ->
                dsl.fetch(
                    "SELECT * FROM alarms WHERE user_id = ? ORDER BY updated_at DESC LIMIT ? OFFSET ?",
                    userId,
                    normalizedPageSize,
                    offset
                )
            userId != null ->
                dsl.fetch(
                    """
                    SELECT * FROM alarms
                    WHERE user_id = ? AND (title LIKE ? OR client_uuid LIKE ?)
                    ORDER BY updated_at DESC
                    LIMIT ? OFFSET ?
                    """.trimIndent(),
                    userId,
                    "%${query!!.trim()}%",
                    "%${query.trim()}%",
                    normalizedPageSize,
                    offset
                )
            query.isNullOrBlank() ->
                dsl.fetch(
                    "SELECT * FROM alarms ORDER BY updated_at DESC LIMIT ? OFFSET ?",
                    normalizedPageSize,
                    offset
                )
            else ->
                dsl.fetch(
                    """
                    SELECT * FROM alarms
                    WHERE title LIKE ? OR client_uuid LIKE ?
                    ORDER BY updated_at DESC
                    LIMIT ? OFFSET ?
                    """.trimIndent(),
                    "%${query.trim()}%",
                    "%${query.trim()}%",
                    normalizedPageSize,
                    offset
                )
        }.map(syncService::mapAlarm)
        val total = when {
            userId != null && query.isNullOrBlank() ->
                (dsl.fetchValue("SELECT COUNT(*) FROM alarms WHERE user_id = ?", userId) as? Number)?.toInt() ?: 0
            userId != null ->
                (dsl.fetchValue(
                    """
                    SELECT COUNT(*)
                    FROM alarms
                    WHERE user_id = ? AND (title LIKE ? OR client_uuid LIKE ?)
                    """.trimIndent(),
                    userId,
                    "%${query!!.trim()}%",
                    "%${query.trim()}%"
                ) as? Number)?.toInt() ?: 0
            query.isNullOrBlank() ->
                (dsl.fetchValue("SELECT COUNT(*) FROM alarms") as? Number)?.toInt() ?: 0
            else ->
                (dsl.fetchValue(
                    """
                    SELECT COUNT(*)
                    FROM alarms
                    WHERE title LIKE ? OR client_uuid LIKE ?
                    """.trimIndent(),
                    "%${query.trim()}%",
                    "%${query.trim()}%"
                ) as? Number)?.toInt() ?: 0
        }
        return PageResponse(
            items = items,
            total = total,
            page = normalizedPage,
            pageSize = normalizedPageSize
        )
    }

    fun getAlarm(alarmId: Long): AlarmSyncDto =
        dsl.fetchOne("SELECT * FROM alarms WHERE id = ? LIMIT 1", alarmId)
            ?.let(syncService::mapAlarm)
            ?: throw ApiException(HttpStatus.NOT_FOUND, "alarm not found")

    fun updateAlarm(adminUserId: Long, alarmId: Long, request: UpdateAlarmRequest, ipAddress: String): AlarmSyncDto = dsl.transactionResult { cfg ->
        val tx = DSL.using(cfg)
        val existing = tx.fetchOne("SELECT * FROM alarms WHERE id = ? LIMIT 1", alarmId)
            ?: throw ApiException(HttpStatus.NOT_FOUND, "alarm not found")
        tx.execute(
            """
            UPDATE alarms SET
              title = ?, note = ?, enabled = ?, status = ?, trigger_time = ?, start_date = ?, end_date = ?,
              schedule_mode = ?, alert_policy = ?, time_anchor_mode = ?, interval_months = ?, interval_years = ?,
              template_id = ?, updated_at = ?
            WHERE id = ?
            """.trimIndent(),
            request.title ?: existing.get("title", String::class.java),
            request.note ?: existing.get("note", String::class.java),
            request.enabled?.let { if (it) 1 else 0 } ?: existing.get("enabled", Int::class.java),
            request.status ?: existing.get("status", Int::class.java),
            request.triggerTime?.toInstantUtc()?.toSqlTimestamp() ?: existing.get("trigger_time", Timestamp::class.java),
            request.startDate?.toInstantUtc()?.toSqlTimestamp() ?: existing.get("start_date", Timestamp::class.java),
            request.endDate?.toInstantUtc()?.toSqlTimestamp() ?: existing.get("end_date", Timestamp::class.java),
            request.scheduleMode ?: existing.get("schedule_mode", Int::class.java),
            request.alertPolicy ?: existing.get("alert_policy", Int::class.java),
            request.timeAnchorMode ?: existing.get("time_anchor_mode", Int::class.java),
            request.intervalMonths ?: existing.get("interval_months", Int::class.java),
            request.intervalYears ?: existing.get("interval_years", Int::class.java),
            request.templateId ?: existing.get("template_id", String::class.java),
            Instant.now(clock).toSqlTimestamp(),
            alarmId
        )
        logAudit(tx, adminUserId, "ALARM_UPDATE", "ALARM", alarmId.toString(), request, ipAddress)
        tx.fetchOne("SELECT * FROM alarms WHERE id = ? LIMIT 1", alarmId)
            ?.let(syncService::mapAlarm)
            ?: throw ApiException(HttpStatus.NOT_FOUND, "alarm not found")
    }

    fun softDeleteAlarm(adminUserId: Long, alarmId: Long, ipAddress: String) = dsl.transaction { cfg ->
        val tx = DSL.using(cfg)
        tx.execute(
            "UPDATE alarms SET status = 1, updated_at = ? WHERE id = ?",
            Instant.now(clock).toSqlTimestamp(),
            alarmId
        )
        logAudit(tx, adminUserId, "ALARM_SOFT_DELETE", "ALARM", alarmId.toString(), mapOf("status" to 1), ipAddress)
    }

    fun listAlarmLogs(userId: Long?, page: Int, pageSize: Int): PageResponse<AlarmLogAdminDto> {
        val normalizedPage = page.coerceAtLeast(1)
        val normalizedPageSize = pageSize.coerceIn(10, 100)
        val offset = (normalizedPage - 1) * normalizedPageSize
        val records = if (userId == null) {
            dsl.fetch(
                "SELECT * FROM alarm_logs ORDER BY fired_at DESC LIMIT ? OFFSET ?",
                normalizedPageSize,
                offset
            )
        } else {
            dsl.fetch(
                "SELECT * FROM alarm_logs WHERE user_id = ? ORDER BY fired_at DESC LIMIT ? OFFSET ?",
                userId,
                normalizedPageSize,
                offset
            )
        }
        val total = if (userId == null) {
            (dsl.fetchValue("SELECT COUNT(*) FROM alarm_logs") as? Number)?.toInt() ?: 0
        } else {
            (dsl.fetchValue("SELECT COUNT(*) FROM alarm_logs WHERE user_id = ?", userId) as? Number)?.toInt() ?: 0
        }
        return PageResponse(
            items = records.map { record ->
            AlarmLogAdminDto(
                id = record.get("id", Long::class.java)!!,
                alarmId = record.get("alarm_id", Long::class.java)!!,
                userId = record.get("user_id", Long::class.java)!!,
                firedAt = record.get("fired_at", Timestamp::class.java)!!.toInstant().toIsoUtc(),
                action = record.get("action", Int::class.java) ?: 0,
                deviceId = record.get("device_id", String::class.java) ?: "",
                logHash = record.get("log_hash", String::class.java) ?: ""
            )
            },
            total = total,
            page = normalizedPage,
            pageSize = normalizedPageSize
        )
    }

    fun listAuditLogs(page: Int, pageSize: Int): PageResponse<AdminAuditLogDto> {
        val normalizedPage = page.coerceAtLeast(1)
        val normalizedPageSize = pageSize.coerceIn(10, 100)
        val offset = (normalizedPage - 1) * normalizedPageSize
        val items = dsl.fetch(
            "SELECT * FROM admin_audit_logs ORDER BY created_at DESC LIMIT ? OFFSET ?",
            normalizedPageSize,
            offset
        ).map { record ->
            AdminAuditLogDto(
                id = record.get("id", Long::class.java)!!,
                adminUserId = record.get("admin_user_id", Long::class.java)!!,
                action = record.get("action", String::class.java) ?: "",
                targetType = record.get("target_type", String::class.java) ?: "",
                targetId = record.get("target_id", String::class.java),
                detailJson = record.get("detail_json", String::class.java),
                ipAddress = record.get("ip_address", String::class.java),
                createdAt = record.get("created_at", Timestamp::class.java)!!.toInstant().toIsoUtc()
            )
        }
        val total = (dsl.fetchValue("SELECT COUNT(*) FROM admin_audit_logs") as? Number)?.toInt() ?: 0
        return PageResponse(
            items = items,
            total = total,
            page = normalizedPage,
            pageSize = normalizedPageSize
        )
    }

    private fun logAudit(
        tx: DSLContext,
        adminUserId: Long,
        action: String,
        targetType: String,
        targetId: String?,
        detail: Any,
        ipAddress: String
    ) {
        tx.execute(
            """
            INSERT INTO admin_audit_logs
            (admin_user_id, action, target_type, target_id, detail_json, ip_address, created_at)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """.trimIndent(),
            adminUserId,
            action,
            targetType,
            targetId,
            objectMapper.writeValueAsString(detail),
            ipAddress,
            Instant.now(clock).toSqlTimestamp()
        )
    }

    private fun mapUser(record: Record): UserProfile = UserProfile(
        id = record.get("id", Long::class.java)!!,
        phone = record.get("phone", String::class.java),
        email = record.get("email", String::class.java),
        nickname = record.get("nickname", String::class.java),
        role = com.smartclock.server.model.Role.valueOf(record.get("role", String::class.java)!!),
        status = record.get("status", Int::class.java) ?: 0
    )
}
